File: platform/packages_py/fetch_httpx/_cache/wrapper/caching_client.py
"""
Caching Client Wrapper

AsyncClient wrapper with built-in caching support.
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Any

from ..._client import AsyncClient
from ..._exceptions import CacheError, CacheWriteError
from ..core import CacheManager
from ..types import CacheConfig, RequestCacheOptions

logger = logging.getLogger("fetch_httpx.cache")

if TYPE_CHECKING:
    from ..._models import Response
    from ..._types import (
        AuthTypes,
        CookieTypes,
        HeaderTypes,
        QueryParamTypes,
        RequestContent,
        RequestData,
        RequestFiles,
        TimeoutTypes,
        URLTypes,
    )


class CachingClient:
    """
    AsyncClient wrapper with caching support.

    Wraps an AsyncClient and adds caching functionality
    for GET/HEAD requests (configurable).

    Example:
        async with CachingClient(
            base_url="https://api.example.com",
            cache=CacheConfig(ttl=60.0)
        ) as client:
            # First request fetches from network
            response = await client.get("/users")

            # Second request returns from cache
            response = await client.get("/users")

            # Manual cache operations
            await client.cache.invalidate("GET:https://api.example.com/users")
            await client.cache.clear()

        # Custom key strategy with dot notation
        from fetch_httpx._cache import create_dot_notation_key_strategy

        key_strategy = create_dot_notation_key_strategy([
            "headers.Authorization",
            "params.page",
        ])

        async with CachingClient(
            base_url="https://api.example.com",
            cache=CacheConfig(ttl=60.0, key_strategy=key_strategy)
        ) as client:
            response = await client.get("/users", params={"page": 1})
    """

    def __init__(
        self,
        cache: CacheConfig | None = None,
        **client_kwargs: Any,
    ) -> None:
        """
        Initialize a caching client.

        Args:
            cache: Cache configuration (ttl, storage, key_strategy, etc.)
            **client_kwargs: Arguments passed to AsyncClient
        """
        self._client = AsyncClient(**client_kwargs)
        self._cache_manager = CacheManager(cache)
        self._closed = False

    @property
    def cache(self) -> CacheManager:
        """Access the cache manager for manual operations."""
        return self._cache_manager

    async def request(
        self,
        method: str,
        url: URLTypes,
        *,
        content: RequestContent | None = None,
        data: RequestData | None = None,
        files: RequestFiles | None = None,
        json: Any | None = None,
        params: QueryParamTypes | None = None,
        headers: HeaderTypes | None = None,
        cookies: CookieTypes | None = None,
        auth: AuthTypes | None = None,
        follow_redirects: bool | None = None,
        timeout: TimeoutTypes | None = None,
        cache_options: RequestCacheOptions | None = None,
    ) -> Response:
        """
        Send an HTTP request with caching support.

        Args:
            method: HTTP method
            url: Request URL
            content: Raw request body
            data: Form data
            files: Files to upload
            json: JSON data
            params: Query parameters
            headers: Request headers
            cookies: Request cookies
            auth: Authentication
            follow_redirects: Follow redirects
            timeout: Request timeout
            cache_options: Per-request cache options

        Returns:
            Response object (from cache or network)
        """
        # Convert to dict for key generation
        headers_dict = dict(headers.items()) if hasattr(headers, "items") and headers else None
        params_dict = dict(params) if params else None
        body_data = json if json is not None else data

        # Build full URL for cache key
        full_url = str(url)
        if self._client._base_url:
            full_url = str(self._client._base_url.join(full_url))

        # Generate cache key
        cache_key = (
            cache_options.cache_key
            if cache_options and cache_options.cache_key
            else self._cache_manager.generate_key(
                method, full_url, headers_dict, body_data, params_dict
            )
        )

        # Check if should cache
        should_cache = self._cache_manager.should_cache(method, cache_options)
        force_refresh = cache_options.force_refresh if cache_options else False

        logger.debug(
            "Cache decision",
            extra={
                "method": method,
                "url": full_url[:100],
                "cache_key": cache_key[:50],
                "should_cache": should_cache,
                "force_refresh": force_refresh,
            },
        )

        # Try to get from cache
        if should_cache and not force_refresh:
            if self._cache_manager.stale_while_revalidate_enabled:
                entry, is_stale = await self._cache_manager.get_stale(cache_key)
                if entry:
                    response = self._cache_manager.create_response_from_cache(entry)
                    if is_stale:
                        # Refresh in background (fire and forget)
                        import asyncio

                        asyncio.create_task(
                            self._fetch_and_cache(
                                method,
                                url,
                                cache_key,
                                cache_options,
                                content=content,
                                data=data,
                                files=files,
                                json=json,
                                params=params,
                                headers=headers,
                                cookies=cookies,
                                auth=auth,
                                follow_redirects=follow_redirects,
                                timeout=timeout,
                            )
                        )
                    return response
            else:
                entry = await self._cache_manager.get(cache_key)
                if entry:
                    return self._cache_manager.create_response_from_cache(entry)

        # Fetch from network
        response = await self._client.request(
            method,
            url,
            content=content,
            data=data,
            files=files,
            json=json,
            params=params,
            headers=headers,
            cookies=cookies,
            auth=auth,
            follow_redirects=follow_redirects,
            timeout=timeout,
        )

        # Cache successful responses
        if should_cache and response.is_success:
            ttl = cache_options.ttl if cache_options else None
            logger.info(
                "Caching response",
                extra={
                    "cache_key": cache_key[:50],
                    "status_code": response.status_code,
                    "ttl": ttl,
                },
            )
            try:
                await self._cache_manager.set(cache_key, response, ttl)
                logger.info(
                    "Response cached successfully",
                    extra={"cache_key": cache_key[:50]},
                )
            except Exception as e:
                logger.error(
                    "Failed to cache response",
                    extra={
                        "cache_key": cache_key[:50],
                        "error": str(e),
                        "error_type": type(e).__name__,
                    },
                )
                raise
        elif should_cache and not response.is_success:
            logger.debug(
                "Not caching - response not successful",
                extra={
                    "cache_key": cache_key[:50],
                    "status_code": response.status_code,
                },
            )
        elif not should_cache:
            logger.debug(
                "Not caching - caching disabled for this request",
                extra={"method": method},
            )

        return response

    async def _fetch_and_cache(
        self,
        method: str,
        url: URLTypes,
        cache_key: str,
        cache_options: RequestCacheOptions | None,
        **kwargs: Any,
    ) -> None:
        """
        Fetch and cache a response (for stale-while-revalidate).

        Errors are logged but not raised since this runs in background.
        """
        try:
            response = await self._client.request(method, url, **kwargs)
            if response.is_success:
                ttl = cache_options.ttl if cache_options else None
                try:
                    await self._cache_manager.set(cache_key, response, ttl)
                except Exception as cache_err:
                    logger.warning(
                        "Background cache update failed",
                        extra={
                            "cache_key": cache_key[:50],
                            "error": str(cache_err),
                            "error_type": type(cache_err).__name__,
                        },
                    )
        except Exception as fetch_err:
            logger.warning(
                "Background cache revalidation failed",
                extra={
                    "url": str(url),
                    "method": method,
                    "cache_key": cache_key[:50],
                    "error": str(fetch_err),
                    "error_type": type(fetch_err).__name__,
                },
            )

    # Convenience methods
    async def get(
        self,
        url: URLTypes,
        *,
        params: QueryParamTypes | None = None,
        headers: HeaderTypes | None = None,
        cookies: CookieTypes | None = None,
        auth: AuthTypes | None = None,
        follow_redirects: bool | None = None,
        timeout: TimeoutTypes | None = None,
        cache_options: RequestCacheOptions | None = None,
    ) -> Response:
        """Send a GET request with caching."""
        return await self.request(
            "GET",
            url,
            params=params,
            headers=headers,
            cookies=cookies,
            auth=auth,
            follow_redirects=follow_redirects,
            timeout=timeout,
            cache_options=cache_options,
        )

    async def head(
        self,
        url: URLTypes,
        *,
        params: QueryParamTypes | None = None,
        headers: HeaderTypes | None = None,
        cookies: CookieTypes | None = None,
        auth: AuthTypes | None = None,
        follow_redirects: bool | None = None,
        timeout: TimeoutTypes | None = None,
        cache_options: RequestCacheOptions | None = None,
    ) -> Response:
        """Send a HEAD request with caching."""
        return await self.request(
            "HEAD",
            url,
            params=params,
            headers=headers,
            cookies=cookies,
            auth=auth,
            follow_redirects=follow_redirects,
            timeout=timeout,
            cache_options=cache_options,
        )

    async def post(
        self,
        url: URLTypes,
        *,
        content: RequestContent | None = None,
        data: RequestData | None = None,
        files: RequestFiles | None = None,
        json: Any | None = None,
        params: QueryParamTypes | None = None,
        headers: HeaderTypes | None = None,
        cookies: CookieTypes | None = None,
        auth: AuthTypes | None = None,
        follow_redirects: bool | None = None,
        timeout: TimeoutTypes | None = None,
        cache_options: RequestCacheOptions | None = None,
    ) -> Response:
        """Send a POST request with optional caching."""
        return await self.request(
            "POST",
            url,
            content=content,
            data=data,
            files=files,
            json=json,
            params=params,
            headers=headers,
            cookies=cookies,
            auth=auth,
            follow_redirects=follow_redirects,
            timeout=timeout,
            cache_options=cache_options,
        )

    async def put(
        self,
        url: URLTypes,
        *,
        content: RequestContent | None = None,
        data: RequestData | None = None,
        files: RequestFiles | None = None,
        json: Any | None = None,
        params: QueryParamTypes | None = None,
        headers: HeaderTypes | None = None,
        cookies: CookieTypes | None = None,
        auth: AuthTypes | None = None,
        follow_redirects: bool | None = None,
        timeout: TimeoutTypes | None = None,
        cache_options: RequestCacheOptions | None = None,
    ) -> Response:
        """Send a PUT request with optional caching."""
        return await self.request(
            "PUT",
            url,
            content=content,
            data=data,
            files=files,
            json=json,
            params=params,
            headers=headers,
            cookies=cookies,
            auth=auth,
            follow_redirects=follow_redirects,
            timeout=timeout,
            cache_options=cache_options,
        )

    async def patch(
        self,
        url: URLTypes,
        *,
        content: RequestContent | None = None,
        data: RequestData | None = None,
        files: RequestFiles | None = None,
        json: Any | None = None,
        params: QueryParamTypes | None = None,
        headers: HeaderTypes | None = None,
        cookies: CookieTypes | None = None,
        auth: AuthTypes | None = None,
        follow_redirects: bool | None = None,
        timeout: TimeoutTypes | None = None,
        cache_options: RequestCacheOptions | None = None,
    ) -> Response:
        """Send a PATCH request with optional caching."""
        return await self.request(
            "PATCH",
            url,
            content=content,
            data=data,
            files=files,
            json=json,
            params=params,
            headers=headers,
            cookies=cookies,
            auth=auth,
            follow_redirects=follow_redirects,
            timeout=timeout,
            cache_options=cache_options,
        )

    async def delete(
        self,
        url: URLTypes,
        *,
        params: QueryParamTypes | None = None,
        headers: HeaderTypes | None = None,
        cookies: CookieTypes | None = None,
        auth: AuthTypes | None = None,
        follow_redirects: bool | None = None,
        timeout: TimeoutTypes | None = None,
        cache_options: RequestCacheOptions | None = None,
    ) -> Response:
        """Send a DELETE request with optional caching."""
        return await self.request(
            "DELETE",
            url,
            params=params,
            headers=headers,
            cookies=cookies,
            auth=auth,
            follow_redirects=follow_redirects,
            timeout=timeout,
            cache_options=cache_options,
        )

    async def aclose(self) -> None:
        """Close the client and cache resources."""
        if self._closed:
            return
        self._closed = True
        await self._cache_manager.close()
        await self._client.aclose()

    async def __aenter__(self) -> CachingClient:
        return self

    async def __aexit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
        await self.aclose()

File: platform/packages_mjs/fetch_undici/src/cache/wrapper/caching-client.ts
/**
 * Caching Client
 *
 * Wrapper class that adds caching to AsyncClient.
 */

import { AsyncClient } from '../../client/client.js'
import type { AsyncClientOptions, RequestOptions } from '../../client/options.js'
import type { Response } from '../../models/response.js'
import { logger, type Logger } from '../../logger.js'
import { CacheManager } from '../core/cache-manager.js'
import type { CacheConfig, RequestCacheOptions } from '../types.js'

const log = logger.create('fetch-undici', import.meta.url)

/**
 * Caching client options
 */
export interface CachingClientOptions extends AsyncClientOptions {
  /** Cache configuration */
  cache?: CacheConfig
}

/**
 * Request options with cache support
 */
export interface CachingRequestOptions extends RequestOptions {
  /** Per-request cache options */
  cache?: RequestCacheOptions
}

/**
 * CachingClient - HTTP client with built-in response caching
 *
 * Wraps AsyncClient and adds transparent caching support.
 *
 * @example
 * ```typescript
 * const client = new CachingClient({
 *   baseUrl: 'https://api.example.com',
 *   cache: { ttl: 60000 }
 * })
 *
 * // First request - cache miss, fetches from network
 * const response1 = await client.get('/users')
 *
 * // Second request - cache hit, returns cached data
 * const response2 = await client.get('/users')
 *
 * // Skip cache for this request
 * const response3 = await client.get('/users', { cache: { noCache: true } })
 *
 * // Force refresh
 * const response4 = await client.get('/users', { cache: { forceRefresh: true } })
 *
 * // Manual cache operations
 * await client.cache.invalidate('GET:https://api.example.com/users')
 * await client.cache.invalidatePattern('GET:https://api.example.com/users/*')
 * await client.cache.clear()
 *
 * await client.close()
 * ```
 */
export class CachingClient {
  private readonly _client: AsyncClient
  private readonly _cacheManager: CacheManager
  private readonly _baseUrl?: string
  private readonly _log: Logger

  constructor(options?: CachingClientOptions) {
    this._log = options?.logger ?? log
    this._baseUrl = options?.baseUrl
    this._client = new AsyncClient(options)
    this._cacheManager = new CacheManager(options?.cache)

    this._log.info('CachingClient created', {
      hasBaseUrl: !!this._baseUrl,
      cacheEnabled: true,
      ttl: options?.cache?.ttl ?? 60000
    })
  }

  /**
   * Access the cache manager for manual operations
   */
  get cache(): CacheManager {
    return this._cacheManager
  }

  /**
   * Access the underlying AsyncClient
   */
  get client(): AsyncClient {
    return this._client
  }

  /**
   * Make HTTP request with caching
   */
  async request(
    method: string,
    url: string,
    options?: CachingRequestOptions
  ): Promise<Response> {
    const cacheOptions = options?.cache

    // Check if caching is applicable
    if (!this._cacheManager.shouldCache(method, cacheOptions)) {
      return this._client.request(method, url, options)
    }

    // Resolve full URL
    const fullUrl = this._resolveUrl(url)

    // Generate cache key
    const cacheKey =
      cacheOptions?.cacheKey ??
      this._cacheManager.generateKey(
        method,
        fullUrl,
        options?.headers as Record<string, string>,
        options?.json ?? options?.data,
        options?.params as Record<string, unknown>
      )

    // Check cache (unless force refresh)
    if (!cacheOptions?.forceRefresh) {
      const cached = await this._cacheManager.get(cacheKey)
      if (cached) {
        this._log.debug('Cache hit', { key: cacheKey, method, url })
        return this._cacheManager.createResponseFromCache(cached)
      }
    }

    // Cache miss - fetch from network
    this._log.debug('Cache miss', { key: cacheKey, method, url })
    const response = await this._client.request(method, url, options)

    // Store in cache if successful
    if (response.ok) {
      await this._cacheManager.set(cacheKey, response, cacheOptions?.ttl)
    }

    return response
  }

  /**
   * GET request
   */
  async get(url: string, options?: CachingRequestOptions): Promise<Response> {
    return this.request('GET', url, options)
  }

  /**
   * POST request
   */
  async post(url: string, options?: CachingRequestOptions): Promise<Response> {
    return this.request('POST', url, options)
  }

  /**
   * PUT request
   */
  async put(url: string, options?: CachingRequestOptions): Promise<Response> {
    return this.request('PUT', url, options)
  }

  /**
   * PATCH request
   */
  async patch(url: string, options?: CachingRequestOptions): Promise<Response> {
    return this.request('PATCH', url, options)
  }

  /**
   * DELETE request
   */
  async delete(url: string, options?: CachingRequestOptions): Promise<Response> {
    return this.request('DELETE', url, options)
  }

  /**
   * HEAD request
   */
  async head(url: string, options?: CachingRequestOptions): Promise<Response> {
    return this.request('HEAD', url, options)
  }

  /**
   * OPTIONS request
   */
  async options(url: string, options?: CachingRequestOptions): Promise<Response> {
    return this.request('OPTIONS', url, options)
  }

  /**
   * Close client and cache
   */
  async close(): Promise<void> {
    await this._cacheManager.close()
    await this._client.close()
    this._log.info('CachingClient closed')
  }

  /**
   * Symbol.asyncDispose for `await using` syntax
   */
  async [Symbol.asyncDispose](): Promise<void> {
    await this.close()
  }

  /**
   * Resolve URL against base URL
   */
  private _resolveUrl(url: string): string {
    if (!this._baseUrl) {
      return url
    }

    // If URL is absolute, return as-is
    if (url.startsWith('http://') || url.startsWith('https://')) {
      return url
    }

    // Join with base URL
    const base = this._baseUrl.endsWith('/') ? this._baseUrl.slice(0, -1) : this._baseUrl
    const path = url.startsWith('/') ? url : `/${url}`
    return `${base}${path}`
  }
}
